### PR TITLE
feat(kotlin-generator): default value for record type fields

### LINKED ISSUE #432: [code generation] Generate record fields default
**Describe the bug**
When generating Kotlin code from avro schemas using KotlinGenerator (directly or using the gradle plugin) the generated Kotlin code is not setting the default value for record fields.

**To Reproduce**
Example schema that gives Kotlin code with lacking feature:

```avsc
{
  "type": "record",
  "name": "Test",
  "fields": [
    {
      "name": "id",
      "type": {
        "type": "record",
        "name": "Id",
        "fields": [
          {
            "name": "name",
            "type": "string",
            "default": ""
          },
          {
            "name": "age",
            "type": "int",
            "default": -1
          }
        ]
      },
      "default": {
        "name": "",
        "age": -1
      }
    }
  ]
}
```

The resulting generated Kotlin code:

Test.kt
```kotlin
@Serializable
@AvroGenerated("""{"type":"record","name":"Test","fields":[{"name":"id","type":{"type":"record","name":"Id","fields":[{"name":"name","type":"string","default":""},{"name":"age","type":"int","default":-1}]},"default":{"name":"","age":-1}}]}""")
public data class Test(
    /**
     * Default value: {name=, age=-1}
     */
    @AvroDefault("{\"name\":\"\",\"age\":-1}")
    public val id: Id,
)
```

Id.kt
```kotlin
@Serializable
@AvroGenerated("""{"type":"record","name":"Id","fields":[{"name":"name","type":"string","default":""},{"name":"age","type":"int","default":-1}]}""")
public data class Id(
    /**
     * Default value: 
     */
    @AvroDefault("")
    public val name: String = "",
    /**
     * Default value: -1
     */
    @AvroDefault("-1")
    public val age: Int = -1,
)
```

**Expected behavior**
I would expect the `id` field to have a default value, such as

ExpectedTest.kt
```kotlin
@Serializable
@AvroGenerated("""{"type":"record","name":"Test","fields":[{"name":"id","type":{"type":"record","name":"Id","fields":[{"name":"name","type":"string","default":""},{"name":"age","type":"int","default":-1}]},"default":{"name":"","age":-1}}]}""")
public data class Test(
    /**
     * Default value: {name=, age=-1}
     */
    @AvroDefault("{\"name\":\"\",\"age\":-1}")
    public val id: Id = Id("", -1),   // Default value for field is defined
)
```

**Desktop:**
 - Kotlin version: 2.3.0
 - Avro4k version: 2.8.0

### PR DESCRIPTION
**Pull request by hermanjakobsen**
Closes #432

Note that the code has been generated by GPT-5.2, due to the author lacking competence in this field of coding. I've tried to include tests that verify the desired behaviour. If this turns out to just be AI slop, I'll happily close the MR. This is my attempt of a quick win, and an experiment to see how efficient AI models can be 😇

### CODE REVIEW

File: kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt [hunk 1]
```diff
@@ -535,8 +522,36 @@ public class KotlinGenerator(
...
-            is TypeSafeSchema.NamedSchema.RecordSchema -> null
+            is TypeSafeSchema.NamedSchema.RecordSchema -> {
+                @Suppress("UNCHECKED_CAST")
+                val defaultMap = (fieldDefault as? Map<*, *>)?.entries?.associate { it.key as String to it.value }
+                    ?: return null
+
+                val args = buildList {
+                    schema.fields.forEach { field ->
+                        val isPresentInDefault = defaultMap.containsKey(field.name)
+                        val valueFromDefault = defaultMap[field.name]
```
Comment (Chuckame) [inline kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt]: Please inline both

File: kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt [hunk 2]
```diff
@@ -535,8 +522,36 @@ public class KotlinGenerator(
...
+
+                        val arg =
+                            if (isPresentInDefault) {
+                                // explicit value provided in default (may be null if nullable)
+                                getRecordFieldDefault(field.schema, valueFromDefault)
+                            } else if (field.hasDefaultValue()) {
+                                // field not present in parent default record; fall back to field's own default
+                                getRecordFieldDefault(field.schema, field.defaultValue)
```
Comment (Chuckame) [inline kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt]: Missing tests

File: kotlin-generator/src/test/resources/nested-record-default-implicit-null.avsc [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        "fields": [
+          {
+            "name": "deep",
+            "type": ["null", {
+              "type": "record",
+              "name": "DeepImplicitNull",
+              "fields": [
+                { "name": "x", "type": "string" }
+              ]
+            }],
```
Comment (Chuckame) [inline kotlin-generator/src/test/resources/nested-record-default-implicit-null.avsc]: Same, not formatted.

File: kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt [hunk 4]
```diff
@@ -554,8 +555,24 @@ public class KotlinGenerator(
...
+                val rawDefaultMap = fieldDefault as? Map<*, *> ?: return null
+                val args = buildList {
+                    schema.fields.forEach { field ->
+                        val valueBlock =
+                            if (rawDefaultMap.containsKey(field.name)) {
+                                getRecordFieldDefault(field.schema, rawDefaultMap[field.name])
+                            } else if (field.hasDefaultValue()) {
+                                getRecordFieldDefault(field.schema, field.defaultValue)
+                            } else {
+                                buildImplicitAvroDefaultCodeBlock(field.schema, implicitNulls = implicitNulls, implicitEmptyCollections = implicitEmptyCollections)
```
Comment (Chuckame) [inline kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt]: Missing unit tests for implicit empty array and implicit empty map

File: kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt [hunk 5]
```diff
@@ -124,15 +125,15 @@ public class KotlinGenerator(
...
-                        } else {
-                            generateNestedKotlinClasses(field.schema, field.name.toPascalCase(), mapOf(schema.asClassName() to recordType))
+                        // generate nested types
+                        schema.fields.flatMap { field ->
+                            // the union schema is already generated as a subtype of the record, no need to generate it again. However, we need to generate the types used in the union
+                            if (field.schema is TypeSafeSchema.UnionSchema) {
+                                field.schema.types.flatMap { generateNestedKotlinClasses(it, potentialAnonymousClassName, emptyMap()) }
+                            } else {
+                                generateNestedKotlinClasses(field.schema, field.name.toPascalCase(), mapOf(schema.asClassName() to recordType))
+                            }
```
[the remark this answers is not in the document]
Comment (hermanjakobsen) [reply kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt]: I'm sorry, my editor went a bit ballistic

File: kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt [hunk 6]
```diff
@@ -535,8 +522,36 @@ public class KotlinGenerator(
                     .takeIf { it.all { it.key is String && it.value != null } }
                     ?.let { getMapOfCodeBlock(it as Map<String, CodeBlock>) }
 
-            // TODO records' defaults are Maps, which needs to be converted to the actual record class instance
-            is TypeSafeSchema.NamedSchema.RecordSchema -> null
+            is TypeSafeSchema.NamedSchema.RecordSchema -> {
+                @Suppress("UNCHECKED_CAST")
+                val defaultMap = (fieldDefault as? Map<*, *>)?.entries?.associate { it.key as String to it.value }
```
[the remark this answers is not in the document]
Comment (hermanjakobsen) [reply kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt]: https://github.com/avro-kotlin/avro4k/pull/434/commits/d34363c2adee2936028fab62c87db22ad35ee99f

File: kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt
Comment (Chuckame) [inline kotlin-generator/src/main/kotlin/com/github/avrokotlin/avro4k/KotlinGenerator.kt]: Here you have a (not that clear) explanation: if the type is not natively serializable (which means only primitive types, enums and data classes goes here), then it ends up to no kotlin default value. During decoding at runtime, it will be handled through the standard way. So here, as logical types are not native types (as they are "logical" so needs post processing), they will fall in the `else` section, which will generate no kotlin default value.

Those kotlin defaults are useful only for the dev xp, where you may not want to pass the parameter for some reason. But it does nothing special regarding decoding or encoding or even performance-wise.

File: kotlin-generator/src/test/resources/nested-record-default-missing-field-default.avsc [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+                    {
+                        "name": "y",
+                        "type": "int",
+                        "default": 42
+                    }
+                ]
+            },
+            "default": {
+                "x": ""
+            }
```
Comment (Chuckame) [inline kotlin-generator/src/test/resources/nested-record-default-missing-field-default.avsc]: Do set an empty default map, and test all the possible types inside this test (at least except record and unions. I don't know if the other field types accept to fallback to the nested field's default)

**Review by Chuckame — CHANGES_REQUESTED** [review]
